### PR TITLE
Combine domain payloads and provide on-the-fly migration

### LINKED ISSUE #3584: Optimize Account Storage Format (reduce domain payloads count by ~29%, etc.)
## Why (Objective)

UPDATE: With Jan 31, 2025 mainnet state, this optimization reduces state size by 11.1%, eliminates 28.2% of mtrie vertexes (leaf and non-leaf nodes), and also reduces payload count by 28.2%.  Our success criteria was to reduce state size by 10-12%, payload count by 25-30%, and eliminate 25-30% of mtrie vertexes. :tada:

Reducing payload count and mtrie vertexes has benefits beyond EN memory reduction.  E.g. last year, when [Atree inlining](https://github.com/onflow/atree/pull/429) in [Atree v0.8](https://github.com/onflow/atree/releases/tag/v0.8.0) reduced payload count by 61% and eliminated 61% of mtrie vertexes, it improved transaction speed by ~7% and @bluesign reported TinyAN bootstrap time reduced by 5x.

We can reduce payload count and mtrie vertex count by:
- replacing all domain payloads per account with one account payload
- using account storage map (atree payload) referencing domain storage map 
 
We can further reduce execution state size by using atree register inlining feature added in [Atree v0.8.0](https://github.com/onflow/atree/releases/tag/v0.8.0).

This improvement can be deployed via HCU and implements first on-the-fly migration to avoid causing further downtime.

The full impact on memory reduction (TBD) will be realized after all the accounts eventually migrate all their non-atree domain registers to a single combined atree payload per account.

NOTE: Memory reduction is only 1 benefit from reducing payload count.

## How will we measure success (Key Results) ?

After all the non-atree domain registers are migrated to 1 atree map per account we expect:
- **reduce execution state size by 10-12% (~40GB as of Sept. 2024 mainnet state)**
- reduce payload count by 20-30% (by combining payloads)
- reduce mtrie nodes by 20-30% (by eliminating mtrie nodes)
- reduce EN memory use by TBD (usually better than state size reduction)

Requires HCU.
If we use on-the-fly migration, then the full impact won't be realized until all accounts are migrated.

Requires "cleanup" migration for accounts that were not migrated on-the-fly.

## Estimate

~3dev/weeks to get ready for deployment.

## DACI

| Role | Assigned |
| ------ | ---------|
| Driver | Technical: @fxamacker , EM: @j1010001|
| Approver | @dete  |
| Consulted| @ramtinms, @zhangchiqing, @turbolent, @janezpodhostnik |
| Informed| Flow engineering team, Product team: @vishalchangrani, Marketing: @SeanRobb  |

## Details

New domains were added in Cadence 1.0 and domain register count increased to 150 million (was 80 million on Sept. 4 pre-spork).  Each 8-byte domain register also requires ~2 mtrie nodes (~2x96 byte overhead).

Nearly 25% of total payloads are 8-byte domain registers.

These 8-byte "domain registers" refers to 8-byte account domain payloads.  They are not atree payloads.

As of Sept 10, 2024, mainnet state had:
- total payload count: 610 million
- atree payload count: 324,800,993
- domain register (aka payload) count: 150,127,970
- etc.

### Suggestion

Currently, all 150+ million account domain registers are not using atree.

I think we should replace non-atree domain registers with 1 atree register (atree map) per account.  Currently, there are 35 million accounts.

Also, we can use atree inlining to inline small domain data.  So instead of reducing payload count from 150 million to 35 million, we can reduce payload count by 174 million.

### UPDATED Estimates (Nov. 2024)

Results based on **full migration test run** using Nov 1, 2024 mainnet state:

- eliminate mtrie nodes: -440 million (-28.8%)
- reduce payload count: -180 million (-28.8%)
- reduce register reads (inlined domain data)
- reduce state size by -43.6GB (-11.4% of state size)
- reduce peak RAM use by TBD (usually better than state size reduction)
- provide single point of entry for account data

We can reduce payload count by ~28% (instead of being limited to ~24%) due to https://github.com/onflow/atree/pull/429.

Full migration test run passed storage health check, diff-states, and used these:
- 2 PRs to combine domain registers
  - https://github.com/onflow/atree/pull/450
  - https://github.com/onflow/cadence/pull/3664
- 3 PRs to reduce state size by an extra -1.2GB (-0.3%) by replacing domain name string with integer:
  - https://github.com/onflow/cadence/pull/3673
  - https://github.com/onflow/cadence/pull/3676
  - https://github.com/onflow/cadence/pull/3677

### DRAFT Estimates (Sept. 2024)

<details><summary>🔍  Initial estimates</summary>

Based on preliminary estimates using Sept. 17, 2024 mainnet state, I think we can:

- eliminate mtrie nodes: -425 million (-28.5%)
- reduce payload count: -174 million (-28.5%)
- reduce register reads (inlined domain data)
- reduce state size by ~40GB (roughly 10-12% of state size)
- reduce peak RAM use by TBD (usually better than state size)
- provide single point of entry for account data

We can reduce payload count by 28% (instead of being limited to 24%) due to [atree inlining](https://github.com/onflow/atree/pull/429).

</details>

### Misc

This issue replaces my older issue #2867.

EDIT: add DRAFT estimates and mention why we can reduce 174 million payload count if there are only 150 million domain registers.

## Effort Estimation

2dev/weeks to wrap-up testing and open PR for review. Will need @turbolent to review.

### Request from marketing:

please write a small Flow forum post (author still TBD, this shouldn't necessarily default onto Faye - though Faye could be an author)
- [x] DRAFT by @fxamacker with 2 items https://www.notion.so/flowfoundation/Account-Storage-Format-v2-and-Zero-Downtime-Migration-forum-post-1911aee123248086bf32d745d6dafd41?pvs=4
  - Account Storage Format v2
    - UPDATE: "Optimized Account Storage Format v2" (thanks @vishalchangrani! 👍) 
  - Zero-Downtime Migrations

### PR DESCRIPTION
**Pull request by fxamacker**
Closes #3584 

## Description

This PR:
1. Combines all domain (non-atree) payloads into one account (non-atree) payload per account.
2. Combines all domain (atree) storage maps into one account (atree) storage map per account.
3. Uses on-the-fly (OTF) migration when accounts are modified (write ops), so this won't change idle or read-only accounts.

NOTE: This requires HCU to deploy. Full impact (e.g. eliminating 425+ million mtrie nodes, etc.) won't be seen until all accounts (including idle accounts with no write ops) are eventually migrated.

### Context

Currently, accounts store data on-chain under pre-defined domains, such as "storage".  Each domain requires domain payload (8-byte non-atree payload) and domain storage map payload (atree payload). Also, each payload requires ~2 mtrie nodes (~2x96 byte overhead).

New domains were added in Cadence 1.0 and domain payloads count increased to 150 million on Sept. 4 (was 80 million pre-spork). Nearly 25% of total payloads on-chain are 8-byte domain payloads. Each account on mainnet has an average of ~4 domain payloads and ~4 domain storage maps.

### Solution

This PR creates 1 account (non-atree) payload and 1 account (atree) storage map per account, eliminating all domain (non-atree) payloads and domain storage maps for that given account.

Based on preliminary estimates using Sept. 17, 2024 mainnet state, this approach can:
- eliminate mtrie nodes: -425 million (-28.5%)
- reduce payload count: -174 million (also -28.5%)

This commit also includes on-the-fly migration so we can see improvements to accounts that have write activity without requiring downtime.  Given this, we won't see the full benefits/impact until all accounts (including idle accounts) are eventually migrated (e.g. using full migration or other means).

### Full Migration Test Results (Nov 4, 2024)

Full migration (on test vm) using this PR produced expected results:
- input: mainnet26 state from Nov 1, 2024
- duration: 7m31s to migrate all accounts after loading state (m1-ultramem-160, nw=60)
- mtrie nodes reduction: ~439 million (-28.7%)
- payload count reduction: ~179 million (also -28.7%)

Storage health check and diff-states on the migrated state did not detect any problems. :tada:

NOTE: On-the-fly migration will not produce full impact until all accounts (including idle and read-only accounts) are eventually migrated.

### Initial Code Review Session (Nov 5, 2024)

We agreed on-the-fly migration is feasible because migrating each account avoids unbounded cost: accounts average 4.3 domains each and we can only migrate a max of 9 domains.

No bugs or showstoppers were found in the PR (yet) but the code can be refactored (e.g. more decoupling to reduce complexity).  I used up ~12 out of 15 devdays to open PR and run full migration test.

Great suggestion by Bastian: it would be OK to replace domain name string with integer.

Bastian also suggested looking into possibility of avoiding HCU by using feature flag.

We agreed to create feature branch for this PR and open additional PRs to feature branch.

### TODO

- [x] @fxamacker: design, implement, open PR (Oct 31, 2024)
- [x] @fxamacker: test full migration using mainnet state (Nov 4, 2024)
  - Storage health check and diff-states did not find any problems. :tada:
- [x] @turbolent and Cadence team: initial design & code review meeting (Nov 5, 2024)
- [x] @turbolent and Cadence team: PR review and feedback

Next:
- [ ] @fxamacker: open PR to replace domain name string with enum (uint64 in shortest form)
  - blocked by master branch refactor https://github.com/onflow/cadence/pull/3673
- [ ] @fxamacker & @turbolent : open PR to increase decoupling (make it easier to reason about the code)
- [ ] @fxamacker & @turbolent: optional but it would be good to look into using feature flag to avoid HCU

______

- [x] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

**Comment by fxamacker** [comment]
To test the impact of this PR, I created a full migration program yesterday and ran it on recent mainnet data (Nov 1, 2024 state).

The full migration program on a test vm produced expected results for reducing number of payloads and mtrie nodes by 28.7% each. 🎉

The plan is to deploy this PR as HCU and use its on-the-fly migration, so we won't see full impact until all accounts are eventually migrated (including idle and read-only accounts).

**Comment by fxamacker** [comment]
@turbolent and Cadence team, many thanks for code review session today!

Next steps (as discussed):
- Create a feature branch and merge this PR into the feature branch (after approval)
- Open new PRs to:
  - replace domain key string to integer in account storage map (great idea @turbolent!)
  - refactor to simplify code in storage (no problems spotted but code is complex)
  - look into maybe using a feature flag (to eliminate the need for HCU)

I will also cleanup and share the private gist I presented in the meeting.

**Review by SupunS — APPROVED** [review]

File: runtime/migrate_domain_registers.go
Comment (turbolent) [inline runtime/migrate_domain_registers.go:114]: We might want to just expose a function that can migrate a single account, and then discuss how to use it.

Currently, at the end of a transaction, all accounts that had writes are migrated, which might take a while. 

Instead, we might want to e.g. only migrate the accounts of the transaction's authorizers.

**Review by turbolent — APPROVED** [review]
Great work! I think we can maybe merge this into the feature branch to start with, and then we can open follow-up PRs as discussed, and finally have a review of the final feature branch when merging it into master

**Pull request merged**
